File: java-to-go-projects/05-port-scanner/Main.java
package scanner;

import java.util.List;

/**
 * CLI entry point for the port scanner.
 *
 * Usage: java Main <host> <startPort> <endPort> [timeout] [threads]
 *
 * Go structure:
 *   main.go              — CLI parsing, calls scanner package
 *   scanner/scanner.go   — Scanner struct, ScanRange method (goroutines + WaitGroup)
 *   scanner/result.go    — ScanResult struct
 *
 * Rust structure:
 *   main.rs              — CLI parsing, calls scanner module
 *   scanner.rs           — scan_range (tokio::spawn + JoinSet)
 *   result.rs            — ScanResult struct
 */
public class Main {

    public static void main(String[] args) {
        if (args.length < 3) {
            System.err.println("Usage: scanner <host> <startPort> <endPort> [timeoutMs] [threads]");
            System.exit(1);
        }

        String host = args[0];
        int startPort = Integer.parseInt(args[1]);
        int endPort = Integer.parseInt(args[2]);
        int timeout = args.length > 3 ? Integer.parseInt(args[3]) : 200;
        int threads = args.length > 4 ? Integer.parseInt(args[4]) : 100;

        System.out.printf("Scanning %s ports %d-%d (timeout=%dms, threads=%d)%n",
            host, startPort, endPort, timeout, threads);

        Scanner scanner = new Scanner(host, timeout, threads);
        long t0 = System.currentTimeMillis();
        List<ScanResult> open = scanner.scanRange(startPort, endPort);
        long elapsed = System.currentTimeMillis() - t0;

        if (open.isEmpty()) {
            System.out.println("No open ports found.");
        } else {
            System.out.println("\nOpen ports:");
            for (ScanResult r : open) {
                System.out.println("  " + r);
            }
        }

        System.out.printf("%nScanned %d ports in %dms%n", endPort - startPort + 1, elapsed);
    }
}

File: java-to-go-projects/05-port-scanner/go/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"

	"port-scanner/scanner"
)

func main() {
	if len(os.Args) < 4 {
		fmt.Fprintf(os.Stderr, "Usage: %s <host> <start-port> <end-port> [timeout-ms] [concurrency]\n", os.Args[0])
		os.Exit(1)
	}

	host := os.Args[1]
	startPort, _ := strconv.Atoi(os.Args[2])
	endPort, _ := strconv.Atoi(os.Args[3])

	timeoutMs := 200
	if len(os.Args) > 4 {
		timeoutMs, _ = strconv.Atoi(os.Args[4])
	}
	concurrency := 100
	if len(os.Args) > 5 {
		concurrency, _ = strconv.Atoi(os.Args[5])
	}

	fmt.Printf("Scanning %s ports %d-%d (timeout=%dms, concurrency=%d)\n",
		host, startPort, endPort, timeoutMs, concurrency)

	s := scanner.New(host, timeoutMs, concurrency)

	t0 := time.Now()
	open := s.ScanRange(startPort, endPort)
	elapsed := time.Since(t0).Milliseconds()

	if len(open) == 0 {
		fmt.Println("No open ports found.")
	} else {
		fmt.Println("\nOpen ports:")
		for _, r := range open {
			fmt.Printf("  %s\n", r)
		}
	}

	fmt.Printf("\nScanned %d ports in %dms\n", endPort-startPort+1, elapsed)
}
